### PR TITLE
[SPARK-48843][FOLLOWUP] Adding more tests that would have failed had the fix ap…

### PR DESCRIPTION
**Pull request by nemanjapetr-db**
…ache/spark#47271 not been sumitted. Prevents regression. Hardens previous tests to catch infinite loop caused by head() and first() in Connect. Adds a test that would have caused infinite loop with BindParameters in Classic.

### What changes were proposed in this pull request?
Adding more tests to https://github.com/apache/spark/pull/47271/commits

### Why are the changes needed?
In the case fix in https://github.com/apache/spark/pull/47271/commits is accidentally rolled back, there newly added tests will catch the regression.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Unittests

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: python/pyspark/sql/tests/connect/test_connect_basic.py [hunk 1]
```diff
@@ -637,6 +638,16 @@ def test_namedargs_with_global_limit(self):
         df2 = self.spark.sql(sqlText, args={"val": 1})
         self.assert_eq(df.toPandas(), df2.toPandas())
 
+        self.assert_eq(df.first()[0], datetime.datetime(2022, 12, 25, 10, 30))
+        self.assert_eq(df.first().date, datetime.datetime(2022, 12, 25, 10, 30))
+        self.assert_eq(df.first()[1], 1)
+        self.assert_eq(df.first().val, 1)
+
+        self.assert_eq(df.head()[0], datetime.datetime(2022, 12, 25, 10, 30))
```
Comment (cloud-fan) [inline python/pyspark/sql/tests/connect/test_connect_basic.py:646]: isn't `df.head` the same as `df.first()`?

**Review by uros-db — APPROVED** [review]
iiuc - `first` is alias for `head`, otherwise lgtm

**Comment by nemanjapetr-db** [comment]
Added three more commits to this PR, but then realized the PR is getting too large. Removed them. Fixed the comment on head()/first() redundancy. Please take another look.

File: python/pyspark/sql/tests/connect/test_connect_basic.py (hunk 1 above)
Comment (nemanjapetr-db) [reply python/pyspark/sql/tests/connect/test_connect_basic.py]: It is. Removed redundancy.

**Comment by cloud-fan** [comment]
The TCS failure is unrelated as it's a test only PR. Thanks, merging to master!

**Pull request closed without merging**